### PR TITLE
[Working Branch] Overture Compose

### PR DESCRIPTION
**Pull request by MitchellShiell**
## About the Composer

- The Overture Composer project aims to provide those interested in gaining hands-on experience using an Overture platform a fast and frictionless local installation option. 

- This work being done in parallel with the [Overture platform guides linked here](https://github.com/overture-stack/website/pull/385)

## Running the Composer

1. Clone repo: 
```
git clone -b develop https://github.com/overture-stack/composer.git
```
2. run compose
```
docker compose up -d
```

> [!IMPORTANT] 
> Make sure you have enough resources allocated to docker. This may be overkill but I am currently using a `CPU limit of 8`, `memory limit of 12gb`, `swap of 4gb`, and `virtual disk limit of 128gb`

### Platform Setup Commits
Setting up and running the Overture Platform

- [x] [A. Database Setup (Postgres + PgAdmin4)](https://github.com/overture-stack/composer/pull/1#issuecomment-2128511388)
- [x] [B. Object Storage (Minio)](https://github.com/overture-stack/composer/pull/1#issuecomment-2130039406)
- [x] [C. OAuth (KeyCloak)](https://github.com/overture-stack/composer/pull/1#issuecomment-2130157244)
- [x] [D. File Management (Song, Score, Kafka, Zookeeper)](https://github.com/overture-stack/composer/pull/1#issuecomment-2133791908)
- [x] [E. Search (Maestro, Elasticsearch, Arranger)](https://github.com/overture-stack/composer/pull/1#issuecomment-2134035219)
- [x] [F. Discovery (Stage)](https://github.com/overture-stack/composer/pull/1#issuecomment-2134075621)

### Data Flow Commits
Testing data input, making fixes and automating mock data seeding

- [x] G. Data submission
- [x] H. Data Display (Arranger Configuration)
- [x] I. Data Download
- [x] J. Automated data seeding

## Issues Log

-  https://github.com/oicr-softeng/keycloak-apikeys/issues/15
- [x] https://github.com/overture-stack/composer/issues/15
- [x] https://github.com/overture-stack/composer/issues/13
- [x] https://github.com/overture-stack/composer/issues/10
- [x] https://github.com/overture-stack/composer/issues/9
- [x] https://github.com/overture-stack/composer/issues/8
- [x] https://github.com/overture-stack/composer/issues/7
- [x] https://github.com/overture-stack/composer/issues/6
- [x] https://github.com/overture-stack/composer/issues/2
- [x] https://github.com/overture-stack/composer/issues/5
- [x] https://github.com/overture-stack/composer/issues/3

## Documentation updates required for Overture.bio

- Song env variables and branch for keycloak 
- Score env variables and branch for keycloak 
- All Keycloak installs should reference the prod setup with linked realm.json for easy setup
- Stage Keycloak Integration

### CODE REVIEW

**Comment by MitchellShiell** [comment]
## This commit includes:

- **Two PostgreSQL services** (`keycloak-db` and `song-db`) using the latest PostgreSQL image, configured with environment variables for user credentials and database names. Each database service uses a separate volume for persistent data storage (Data will persist when containers are shut down and restarted).

- **A pgAdmin4 service** for managing the databases, configured with environment variables for login credentials and exposed on port 5051 of the host machine.
    -  With this setup when connecting databases into pgAdmin the host is the container name `keycloak-db` or `song-db`, username/password is `admin`/`admin123` and the port is the default `5432` for both.


>[!NOTE] 
> Postgres databases do not have any exposed ports. All services here are connected to a custom network named `db-network` to isolate communication between services that need to interface with the databases (Song, Score, and Maestro). I believe this is standard practice to ensure databases on the back end are in an isolated and secure environment. Link to [dockers documentation on networks top-level elements here](https://docs.docker.com/compose/compose-file/06-networks/)

**Comment by MitchellShiell** [comment]
## The OAuth (KeyCloak) commit includes:

- **A production image of Keycloak** using the Keycloak Docker image version 22.0 from `quay.io/keycloak/keycloak`. 
    - It sets up **environment variables** for Keycloak admin credentials (`KEYCLOAK_ADMIN` as `admin` and `KEYCLOAK_ADMIN_PASSWORD` as `admin123`), the hostname is `localhost`.
    - **Volumes**: It mounts various configuration files (`server.keystore`, `keycloak-apikeys-1.0.1.jar`, `myrealm-realm.json`, `myrealm-users-0.json`) that have been added and organized in **configurationFiles>keycloakConfigs**
    - **Networks**: It connects to the `db-network` described in the [database commit comment](https://github.com/overture-stack/composer/pull/1#issuecomment-2128511388)
    - **Commands**: It starts Keycloak with additional configurations for importing the realms, specifying PostgreSQL as the database backend, setting the database URL, username, password, HTTPS key store file, and password.
 
    
- Once running you should be able to access the Keycloak admin console from `https://localhost:8443/admin` (make sure to use https) The self-signed certificate included will cause the browser to give a warning. Select **Advanced** and, depending on your browser, click the button to accept the risk and continue.

- The following commits should see several updates to the realm.json files, this will reduce any need for users to configure Keycloak on startup

> [!Note]
> I've set this up so keycloak `depends_on` the keycloak_db `condition` to equal `service_healthy` before running.
> - I added a `healthcheck` to both `keycloak-db` to test the health of the container by running the following command on an interval`pg_isready -U admin -d keycloakDb`. Leo mentioned this as a potential solution and it has worked nicely so far. This prevents Keycloak from exiting on startup if the database is not ready and the connection gets refused. 
> - I have preemptively added the same healthcheck to `song-db`, and will likely need to do the same for `minio`.

**Comment by MitchellShiell** [comment]
## The File Management (Song + Score) commit includes:

- **Dependent services** Kafka and ZooKeeper with healthchecks. 
    - For zookeeper this is `CMD-SHELL", "echo srvr | nc 127.0.0.1 2181 | grep Mode` if mode is returned in the response you can broadly assume it is running. 
    - For Kafka I am listing topics, If kafka is not ready it will provide a warning and exit with a non-zero status (indicating a failure).

- **Song and Score:** Both song and score are running however I need to now need to work with Keycloak to make sure authorization keys (JWT/API keys) work out the gate. Essentially configure the realms imported into keycloak on startup

### Add to To Do:

- Need to consult Leo on the following:
    - Setting up realm configuration for composer
    - Exporting realm configuration as he did before
    - Getting API keys to work for Song and Score 

> [!NOTE] 
> **SPRING_SECURITY_OAUTH2_RESOURCESERVER_JWT_JWK_SET_URI**
> 1. Make sure you are using the keycloak-integration branch
> 2. Make sure to reference the local host at HTTPS if using a production image of keycloak
> 3. ENV variable should be `SPRING_SECURITY_OAUTH2_RESOURCESERVER_JWT_JWK_SET_URI` need to update the documentation, figured this out by referencing the [song-server application.yaml](https://github.com/overture-stack/SONG/blob/keycloak-integration/song-server/src/main/resources/application.yml#L198C1-L209C86)


> [!CAUTION] 
> **Healthchecks for Keycloak & Minio**
> - I wanted to make sure Keycloak and Minio were healthy before spinning up Song and Score (testing from my machine this has not been an issue; however, having healthchecks would cover all potential scenarios)
>   - Curl commands are not natively installed in the Keycloak and Minio docker images however, to do health check tests in the docker compose both require a curl command pinging their health endpoint(s)
>   - For keycloak this is `curl --head -fsS --insecure https://keycloak:8443/health/ready` [Link to documentation](https://www.keycloak.org/server/health)
>   - For Minio this is `curl -I http://localhost:9001/minio/health/live` [Link to documentation](https://min.io/docs/minio/linux/operations/monitoring/healthcheck-probe.html?utm_term=&utm_campaign=&utm_source=adwords&utm_medium=ppc&hsa_acc=8976569894&hsa_cam=20593618271&hsa_grp=&hsa_ad=&hsa_src=x&hsa_tgt=&hsa_kw=&hsa_mt=&hsa_net=adwords&hsa_ver=3&gad_source=1&gclid=Cj0KCQjwmMayBhDuARIsAM9HM8fmw7at99A1DEJNZFIVfa492JyjeD51yCdxPJsTiYerRhAFfBVF4YkaAkNBEALw_wcB)
> - I cannot find a solution that does not require (a) building a custom image from a docker file or (b) running a shell script outside of docker. apt install command from the docker CMD SHELL has not worked for me either

**Comment by MitchellShiell** [comment]
## The Search (Maestro + Elasticsearch + Arranger server) commit includes:

- **Elasticsearch,** much of this is taken from the existing [docker compose](https://github.com/overture-stack/arranger/blob/develop/docker-compose.yml#L4-L26) found in arranger

- **Arranger Server**, here it spins up once elasticsearch is considered healthy. It will typically fail to connect 4 times before elasticsearch is actually ready. Linking [Overture Demo Infra mappings and configs](https://github.com/oicr-softeng/overture-infra/tree/demo/envs/prod/stateless/arranger) for future reference. 

- **Maestro** is setup to spin up last. Referencing [the demo infra values.yml](https://github.com/oicr-softeng/overture-infra/blob/demo/envs/prod/stateless/maestro/values.yml), I believe there are a few env variables to be added such as `MAESTRO_ELASTICSEARCH_INDEXES_FILECENTRIC_ENABLED`,  `MANAGEMENT_HEALTH_ELASTICSEARCH_ENABLED`, etc.. I will look to this when ready to test functionality.

> [!NOTE] 
> **Index & Index Mapping**
> - I have worked out the main directory of interest here is `/app/docker/elasticsearch`
> - Will need some consultation on how to update/seed the index mapping and index, I tried at repurposing the load-es-data.sh script to no-avail, even trying to run it within the container.

**Comment by MitchellShiell** [comment]
# The Discovery (Stage) commit includes:

- **Stage image 297e6eb** will need to create a dedicated image for the composer

    - Also need to update documentation on Overture.bio for integrating keycloak with stage, here is the [internal reference doc](https://wiki.oicr.on.ca/display/OV/DMS+EGO-Keycloak+configuration) I used to integrate Keycloak
    - Will test login functionality, after configuring stage in keycloak, will add this to the list of configurations needed to be included within the realm.json imported on keycloaks startup

**Comment by MitchellShiell** [comment]
# The above fixes commits includes:

### Automated seeding of elasticsearch index & arranger configurations

I configured two Docker containers:

1. **Elasticsearch Container**:
   - I enabled managing index templates (`MANAGE_INDEX_TEMPLATES: "true"`).
   - Everything else is pretty standard and pulled from [the arranger repo docker compose](https://github.com/overture-stack/arranger/blob/develop/docker-compose.yml)

2. **Index Setup Container**:
   - Linux image
   - Mounts configuration files for the Elasticsearch index and component templates that are stored in the repo
   - Composable index templates were introduced in Elasticsearch 7.8, here is the link to elasticsearch's documentation on [Index templates](https://www.elastic.co/guide/en/elasticsearch/reference/current/index-templates.html)
   - Executes a series of commands to install `curl`, wait for `30s`, and then create Elasticsearch component and index templates using `curl`. It also creates an alias for the index.
   - The original index mapping was taken from the demo portal and can be [found here](https://github.com/oicr-softeng/overture-infra/blob/demo/envs/prod/stateless/arranger/mappings/analysis_centric.json)
   - I have renamed it as `analysis-composer-index` here, anything with `analysis-*` will be assigned the index mapping defined within the composer_index_template.json found in the elasticsearchConfigs folder
   - The composer_component_template seems redundant but it contains templated settings and is required with elasticsearch's current templated indexing workflow 

> [!NOTE] 
> The alternative of running these scripts within the elasticsearch image causes the es image to exit on completion. This setup is designed to run Elasticsearch and manage index templates through another container that sets up these templates once Elasticsearch is ready. This solution allows us to set up the index while keeping the elasticsearch running. 

**Arranger Server**

- The server conveniently starts up after the index-setup image exits (on success)
- All configuration files are pulled from the configs stored in the repo and referenced from the demo portal [located here](https://github.com/oicr-softeng/overture-infra/tree/demo/envs/prod/stateless/arranger/configs) 

### Stage + Keycloak updates

- Following this [Stage PR](https://github.com/overture-stack/stage/pull/202), we can now generate API keys from Stage using Keycloak
- Updated the keycloak realm with a default admin user with admin privileges and exported the realm.json and user.json. This now gets automatically configured on the startup of the image

### Maestro Kafka + Zookeeper updates

- A myriad of configuration changes, primarily with maestro's env variables
- maestros [application.yml](https://github.com/overture-stack/maestro/blob/master/maestro-app/src/main/resources/config/application.yml), [docker-compose](https://github.com/overture-stack/maestro/blob/master/run/docker-compose/docker-compose.yml) and the [demo portal values.yml](https://github.com/oicr-softeng/overture-infra/blob/demo/envs/prod/stateless/maestro/values.yml) were useful resources in troubleshooting
- This medium article on [running a Kafka Cluster with Docker Compose ](https://medium.com/@erkndmrl/kafka-cluster-with-docker-compose-5864d50f677e) was somewhat useful to reference too
- using confluentinc/cp-kafka:5.4.0 and confluentinc/cp-zookeeper:latest however i'd like to find the specific image for zookeeper instead of using the latest tag (avoid issues down the line with updates to the image)

**Comment by MitchellShiell** [comment]
# The above commit includes:

### Song Database Initialization with flyaway

Related to issue #7, The following was required to get songs Db initialized correctly:

- `SPRING_FLYWAY_ENABLED`needed to be set to `true`, here is an [article that explains flyaway](https://www.j-labs.pl/en/tech-blog/flyway-migrations-with-spring/#:~:text=Flyway%20is%20an%20open%2Dsource,structure%2C%20like%20hibernate%20auto%20create.)
- This does a database migration using [these SQL scripts found in Song](https://github.com/overture-stack/SONG/tree/develop/song-server/src/main/resources/db/migration)
- This migration sets up the tables we need to when sending information to Song through any of our API endpoint
- Without it we have a blank database which, when contacted, will return a generic SQL error (SQL Error: 0) with a [SQLState of 42P01 ](https://www.postgresql.org/docs/16/errcodes-appendix.html#:~:text=undefined_function-,42P01,-undefined_table), corresponding to an undefined_table

> [!NOTE]
> One odd caveat is that [v1.2 for dynamic schema integration is hardcoded in JAVA here](https://github.com/overture-stack/SONG/blob/develop/song-server/src/main/java/db/migration/V1_2__dynamic_schema_integration.java)

### Configuration Caveat

Previously I have the data source connection url as such:
`SPRING_DATASOURCE_URL: jdbc:postgresql://song-db:5432/songDb`

> [!IMPORTANT] 
> In our setup y**ou absolutely need to append the connection url with** `?stringtype=unspecified`. This is because Song as it is coded requires string type to be unspecified to interact with JSONb columns 

Altogether the corrected configuration is:
`SPRING_DATASOURCE_URL: jdbc:postgresql://song-db:5432/songDb?stringtype=unspecified`

Additionally, Song requires postgres version 11.1

### Running Keycloak over HTTP instead of HTTPS

- Running Keycloak in production mode requires a self signed cert which did not play nicely with the other Overture services running on HTTP
    - This was initially chosen as it was thought as the only way to run Keycloak with its own db
- Turns out running Keycloak in development mode still allows us to connect it to a seperate postgres database
- Running on http eliminates the need for added development on song and score (to accept self signed certs)

**Comment by MitchellShiell** [comment]
# The above commit includes


## Updated Keycloak Admin Configuration:

- Created a single client containing both Song and Score called `dataManagementServices` 

- I've exported and updated the realm.json and [user.json accordingly ](https://github.com/overture-stack/composer/blob/48859073683460d6ae103084251b52270ddb2930/configurationFiles/keycloakConfigs/myrealm-users-0.json)

## Updated ENVs

- **Stage:** `NEXT_PUBLIC_KEYCLOAK_PERMISSION_AUDIENCE:` updated with the value `dataManagementServices` this means stage generates API keys with both Song and Score permissions that exist within the client as a resource

- **Song and Score:** updated AUTH_SERVER_CLIENTID and AUTH_SERVER_CLIENTSECRET in line with above

- Updated Score `S3_ENDPOINT` from `localhost` to `host.docker.internal`

# Related Issue

https://github.com/oicr-softeng/keycloak-apikeys/issues/15

- Although this configuration allows stage now to generate API keys with appropriate permission, we are now unable to generate API keys with additional study permissions as API key values are limited to 255 characters. 

- This means Stage can generate a key with song and score permissions, but not with any additional study permissions.

- API keys permissions are defined as attributes given to the user within the keycloak console (pre-configured using the realm-user.jsons)

```json
{"name":"F4C094A60BA88FB3F42BB9D20D75931286549D7F3C2E448F62D81CE20237B9BC","scope":["score.READ","score.WRITE","song.READ","song.WRITE"],"expiryDate":1830297599000,"issueDate":1701960265122,"isRevoked":false,"description":"System wide ApiKey "}
```

The above will allow for API key generation however if I add any additional resources such as new studies to the `dataManagementServices` these API keys fail to generate.

**Comment by MitchellShiell** [comment]
# The Above Commits Include: 

- **Updated Realm.json** with one client called `dms` (data management services) that includes song, score and a study named SECA
    - Additionally one client `ADMIN` with system wide permissions

- **Updated Song volume:** I added  `.gitkeeps` to make sure the needed empty directory are tracked by git, this creates two error/warning messages in  the song-db startup logs but no issues found as of yet.
    -  I also included a zip folder with the original populated song-db, no .gitkeeps, 
    - both will allow you to spin up the compose with a study `SECA` created, a schema called `composerSchema` submitted and an analysis with the ID `e83a95c3-e2b9-43c3-ba95-c3e2b9e3c338` ready for publish

- **Updated Minio Automation** to create the needed heliograph file if not present on startup

File: docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,476 @@
...
+      SPRING_FLYWAY_ENABLED: true
+      # Song Variables
+      ID_USELOCAL: true
+      SCHEMAS_ENFORCELATEST: true
+      # Score Variables
+      SCORE_URL: http://localhost:8087
+      SCORE_CLIENTCREDENTIALS_ID: score
+      SCORE_CLIENTCREDENTIALS_SECRET: scoresecret
+      SCORE_CLIENTCREDENTIALS_TOKENURL: http://score:8087/oauth/token
+      SCORE_CLIENTCREDENTIALS_SYSTEMSCOPE: score.WRITE
```
Comment (leoraba) [inline docker-compose.yml:168]: comment out this variables. instead use:
`SCORE_ACCESSTOKEN: [INSERT_TOKEN_HERE]`

File: docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,476 @@
...
+        condition: service_healthy
+      kafka:
+        condition: service_healthy
+      zookeeper:
+        condition: service_healthy
+    ports:
+      - "8080:8080"
+    environment:
+      # Spring Variables
+      SPRING_PROFILES_ACTIVE: prod,secure,score-client-cred,kafka
```
Comment (leoraba) [inline docker-compose.yml:157]: Remove `score-client-cre` profile to allow use score access Token Authentication

**Comment by leoraba** [comment]
Alternative solution: Configure `song` with `SCORE_ACCESSTOKEN` instead of client credentials. 
Steps:
1. Make sure to set song profiles without `score-client-cred`
2. Provide a score token `SCORE_ACCESSTOKEN: [INSERT_TOKEN_HERE]`
Use an APIKEY from an user with admin permissions.

**Investigation ongoing:**
*How to set permission to the client `dms`.*
Explanation: When using Client Configuration is required to set token URL to look for keycloak server and a scope as `profile` to authenticate dms client and get a JWT. With the current keycloak configuration the client `dms` is not granted any permission.

**Pull request merged**